File: backend-challenge/src/services/CouponService.ts
import { CouponValidationResult } from '../domain/Coupon';
import { ICouponRepository } from '../repositories/interfaces/ICouponRepository';

const MIN_LENGTH = 8;
const MAX_LENGTH = 10;

export class CouponService {
  constructor(private readonly couponRepository: ICouponRepository) {}

  validate(code: string): CouponValidationResult {
    if (code.length < MIN_LENGTH) {
      return { valid: false, reason: `Coupon code must be at least ${MIN_LENGTH} characters` };
    }

    if (code.length > MAX_LENGTH) {
      return { valid: false, reason: `Coupon code must be at most ${MAX_LENGTH} characters` };
    }

    if (!this.couponRepository.isValid(code)) {
      return { valid: false, reason: 'Coupon code is not valid' };
    }

    return { valid: true };
  }
}

File: go-backend/internal/service/coupon_service.go
package service

import (
	"github.com/nidhey27/kart-challenge/internal/domain"
	"github.com/nidhey27/kart-challenge/internal/repository"
)

type couponService struct {
	repo repository.CouponRepository
}

// NewCouponService returns a CouponService backed by the given repository.
func NewCouponService(repo repository.CouponRepository) CouponService {
	return &couponService{repo: repo}
}

// Validate checks length constraints (8-10 chars) then delegates to the repo.
func (s *couponService) Validate(code string) *domain.CouponValidationResult {
	l := len(code)
	if l < 8 {
		return &domain.CouponValidationResult{Valid: false, Reason: "coupon code too short"}
	}
	if l > 10 {
		return &domain.CouponValidationResult{Valid: false, Reason: "coupon code too long"}
	}
	if !s.repo.IsValid(code) {
		return &domain.CouponValidationResult{Valid: false, Reason: "coupon code not found"}
	}
	return &domain.CouponValidationResult{Valid: true}
}
